pkgs/operator/state.go

```diff
@@ -0,0 +1,280 @@
+	pkBytes, err := crypto.EncodePublicKey(operatorPubKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, op := range init.Operators {
+		if bytes.Equal(op.PubKey, pkBytes) {
+			operatorID = op.ID
+			break
+		}
+	}
+
+	if operatorID == 0 {
+		return nil, nil, errors.New("my operator is missing inside the operators list at instance")
+	}
+
+	bchan := make(chan []byte, 1)
+
+	broadcast := func(msg []byte) error {
+		bchan <- msg
+		return nil
+	}
+
+	opts := dkg.OwnerOpts{
+		Logger:      s.Logger.With(zap.String("instance", hex.EncodeToString(reqID[:]))),
+		BroadcastF:  broadcast,
+		SignFunc:    s.Sign,
+		VerifyFunc:  verify,
+		EncryptFunc: s.Encrypt,
+		DecryptFunc: s.Decrypt,
+		Suite:       bls3.NewBLS12381Suite(),
+		ID:          operatorID,
+		RSAPub:      &s.PrivateKey.PublicKey,
+		Owner:       init.Owner,
+		Nonce:       init.Nonce,
+	}
+	owner := dkg.New(opts)
+	// wait for exchange msg
+	resp, err := owner.Init(reqID, init)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := owner.Broadcast(resp); err != nil {
+		return nil, nil, err
+	}
+	res := <-bchan
+	return &instWrapper{owner, initiatorPublicKey, bchan, owner.ErrorChan}, res, nil
+}
+
+func (s *Switch) Sign(msg []byte) ([]byte, error) {
+	return crypto.SignRSA(s.PrivateKey, msg)
+}
+
+func (s *Switch) Encrypt(msg []byte) ([]byte, error) {
+	return rsa.EncryptPKCS1v15(rand.Reader, &s.PrivateKey.PublicKey, msg)
+}
+
+func (s *Switch) Decrypt(ciphertext []byte) ([]byte, error) {
+	return rsaencryption.DecodeKey(s.PrivateKey, ciphertext)
+}
+func (s *Switch) CreateVerifyFunc(ops []*wire.Operator) (func(id uint64, msg []byte, sig []byte) error, error) {
+	inst_ops := make(map[uint64]*rsa.PublicKey)
+	for _, op := range ops {
+		pk, err := crypto.ParseRSAPubkey(op.PubKey)
+		if err != nil {
+			return nil, err
+		}
+		inst_ops[op.ID] = pk
+	}
+	return func(id uint64, msg []byte, sig []byte) error {
+		pk, ok := inst_ops[id]
+		if !ok {
+			return fmt.Errorf("cant find operator participating at DKG %d", id)
+		}
+		return crypto.VerifyRSA(pk, msg, sig)
+	}, nil
+}
+
+type Switch struct {
+	Logger           *zap.Logger
+	Mtx              sync.RWMutex
+	InstanceInitTime map[InstanceID]time.Time
+	Instances        map[InstanceID]Instance
+	PrivateKey       *rsa.PrivateKey
+}
+
+func NewSwitch(pv *rsa.PrivateKey, logger *zap.Logger) *Switch {
+	return &Switch{
+		Logger:           logger,
+		Mtx:              sync.RWMutex{},
+		InstanceInitTime: make(map[InstanceID]time.Time, MaxInstances),
+		Instances:        make(map[InstanceID]Instance, MaxInstances),
+		PrivateKey:       pv,
+	}
+}
+
+func (s *Switch) InitInstance(reqID [24]byte, initMsg *wire.Transport, initiatorSignature []byte) ([]byte, error) {
+	logger := s.Logger.With(zap.String("reqid", hex.EncodeToString(reqID[:])))
+	logger.Info("🚀 Initializing DKG instance")
+	init := &wire.Init{}
+	if err := init.UnmarshalSSZ(initMsg.Data); err != nil {
+		return nil, err
```

Comment:
We should return error wrappings (as you did below in `VerifyRSA`) here and in the rest of the code, so we can tell where exactly things failed in retrospect.

Reply:
added error wrapps at state. Should we have wrapps everywhere?